### ISSUE TITLE
Discovery error

### ISSUE THREAD

**Issue by prabhudavidsheryl** [openark/orchestrator#520]
Hello,

Just completed an Orchestrator installation(3 instances) with raft to manage a single master 2 slaves  MySQL replication setup. While raft communication seems to work fine the discovery of the Slaves after Master discovery fails.

Error: 

[mysql] 2018/05/30 14:53:45 packets.go:36: read tcp x.x.x.x:34628->x.x.x.x:3306: i/o timeout
2018-05-30 14:53:45 ERROR invalid connection
[mysql] 2018/05/30 14:53:45 packets.go:36: read tcp x.x.x.x:59936->x.x.x.x:3306: i/o timeout
2018-05-30 14:53:45 ERROR invalid connection

MySQL replication and MySQL connection with the mysql client between servers works though.

Setup details:

OS: Centos 7
Orchestrator: 3.0.11
MySQL: 5.6

Any help pointing to known issues would be useful.

**Comment by shlomi-noach** [comment]
1. If I understand correctly, `orchestrator` is able to discover the master but not the two replicas?

2. What do you see if you run `orchestrator-client -c all-instances`, and `orchestrator-client -c topology -i name.of.the.master:3306` ?

3. Are replicas running on port `:3306`?

4. Do you use `DiscoverByShowSlaveHosts` as `true` or `false` in config file?

5. Actually, please paste your config file. Is it identical on all boxes?

6. Are master & replicas all on same network?

7. Do they have same credentials?

**Comment by prabhudavidsheryl** [comment]
Thanks for the quick help!! Answers below:

> If I understand correctly, orchestrator is able to discover the master but not the two replicas? 

Yes

> What do you see if you run orchestrator-client -c all-instances, and orchestrator-client -c topology -i name.of.the.master:3306 ?

```
[root@xxxx-dn8 app]# orchestrator-client -c all-instances
xxxx-dn8:3306


[root@xxxx-dn8 app]# orchestrator-client -c topology -i xxxx-dn8:3306
xxxx-dn8:3306 [0s,ok,5.6.24-log,rw,ROW,>>,GTID]
```

> Are replicas running on port :3306?

Yes

> Do you use DiscoverByShowSlaveHosts as true or false in config file? 

false

> Actually, please paste your config file. Is it identical on all boxes?

> Are master & replicas all on same network? 

Yes

> Do they have same credentials? 

Yes

We tried excluding one node suspecting a server/network issue, but in the 2 node(Orchestrator) setup also we have the same error. Configuration file and Orchestrator log attached.

[orchestrator.zip](https://github.com/github/orchestrator/files/2058038/orchestrator.zip)

**Comment by shlomi-noach** [comment]
Pro tip: see how I've formatted your response, making it more readable.

**Comment by shlomi-noach** [comment]
I see in your log evidence to serious networking errors. This is mainly observed in the time it takes to the raft cluster to setup itself.

But I also notice your `orchestrator` runs on the very same box as the `mysql` master. First, this is discouraged, you want to run `orchestrator` externally, but regardless, this makes more sense in explaining why `orchestrator` is able to connect to the master and not to the replicas. Perhaps your mysql binds to `127.0.0.1`?
Either way, seems like `orchestrator` is able to determine that the replica is `xxxx-mn4:3306` and is then unable to connect to it.

Please check all things network. Try to connect to your replica from the master box using normal `mysql` client.

**Comment by shlomi-noach** [comment]
Also, a `raft` cluster should have at least 3 nodes. It's almost meaningless to have 2 nodes.

**Comment by prabhudavidsheryl** [comment]
Thanks!!

We initially setup a 3 node setup with 1 Orchestrator process on each node with raft and each of these nodes also has 2 MySQL instances, one listening on the default port 3306(The Orchestrator managed setup) and one listening on 3307 to be used by Orchestrator as its repository.

We are in "proof of concept" phase so Orchestrator shares same servers as the instances it manages, we may go with separate servers later.

The logs I shared are from a 2 node setup we tried afterwards following the initial 3 node setup suspecting network issues with one of the servers, but that does not seem to be the case as the 2 node setup also shows the same issue. We will try to get the Network team to check soon.

The confusing part though is that `mysql `client connections works, and so does replication:

```
[root@xxxx-dn8 bandarc1]# mysql -u repadmin -h xxxx-mn4 -p
Enter password:
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 385
Server version: 5.6.24-log MySQL Community Server (GPL)

Copyright (c) 2000, 2015, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql>
```

**Comment by shlomi-noach** [comment]
Is orchestrator using the same credentials as you're using with your MySQL client?

**Comment by prabhudavidsheryl** [comment]
In the test above I used a different account to connect but `orchestrator ` is configured to use an account called `"orchestrator"` and even that uses the same credentials as `"repadmin"`. Its common for all accounts for the timebeing.

**Comment by prabhudavidsheryl** [comment]
```
[root@xxxx-dn8 bandarc1]# mysql -u orchestrator -h xxxx-mn4 -p
Enter password:
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 628
Server version: 5.6.24-log MySQL Community Server (GPL)

Copyright (c) 2000, 2015, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql>
mysql>
mysql> show databases;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| mysql              |
+--------------------+
2 rows in set (0.00 sec)

mysql>
```

**Comment by prabhudavidsheryl** [comment]
Hi,

Have been doing some investigation on this after setting up `Orchestrator ` in standalone mode and on a separate server and configured to use MySQL on the default port.

```
Current setup:

MySQL topology:
xxxx-dn8 Master
xxxx-mn4 Slave

Orchestrator:
xxxx-mn3 
```

On discovery we are getting the same error:

```
[root@xxxx-mn3 orchestrator]# orchestrator-client -c discover -i xxxx-dn8:3306
invalid connection
```

But when discovery is triggered we can see connections in `xxxx-dn8:3306` and the general query log on `xxxx-dn8:3306` shows statements from Orchestrator on xxxx-mn3:

```
180604 13:22:14 16617 Query     show variables like 'maxscale%'
                16617 Query     show global status like 'Uptime'
                16621 Query     select @@global.hostname, ifnull(@@global.report_host, ''), @@global.server_id, @@global.version, @@global.version_comment, @@global.read_only, @@global.binlog_format, @@global.log_bin, @@global.log_slave_updates
                16621 Query     show master status
                16617 Query     show global status like 'rpl_semi_sync_%_status'
                16621 Query     show slave status
                16622 Connect   orchestrator@xxxx-mn3.ref.dab.02.net on
                16622 Query     select @@global.gtid_mode, @@global.server_uuid, @@global.gtid_purged, @@global.master_info_repository = 'TABLE', @@global.binlog_row_image
                16617 Query     select count(*) > 0 and MAX(User_name) != '' from mysql.slave_master_info
180604 13:22:24 16621 Quit
```
This does not look like a network problem as we can see connections from Orchestrator to the topology mysql instances. Can this be a configuration issue please?

Can see `"ERROR FlushNontrivialResolveCacheToDatabase() called, but HostnameResolveMethod is default"` in the log, can this mean anything please.

```
[martini] Started GET /api/discover/xxxx-dn8/3306 for 10.29.2.159:42014
[mysql] 2018/06/04 13:16:10 connection.go:372: invalid connection
[mysql] 2018/06/04 13:16:20 packets.go:36: read tcp 10.29.2.159:58092->10.29.2.141:3306: i/o timeout
2018-06-04 13:16:20 ERROR invalid connection
[martini] Completed 500 Internal Server Error in 10.003636292s
2018-06-04 13:16:47 ERROR FlushNontrivialResolveCacheToDatabase() called, but HostnameResolveMethod is default
```
Regards
Prabhu

**Comment by shlomi-noach** [comment]
`FlushNontrivialResolveCacheToDatabase` is reported in https://github.com/github/orchestrator/issues/515, and is irrelevant to your situation.

My next conjecture would be that, on your master, a `SHOW PROCESSLIST` shows the replica hostnames in a different way than the way you use to connect directly to those replicas.

For example, `SHOW PROCESSLIST` would give the `ip` address, which for some reason` orchestrator` is unable to follow; but maybe when you discover directly you do not pass the `ip` address, but instead provide a `fqdn`?

Or perhaps `SHOW PROCESSLIST` would give you a short hostname (e.g. `xxxx`) whereas when you ask `orchestrator` to discover directly you indicate `xxxx-dn8.mycompany.com`.

Can you please look into that?

**Comment by prabhudavidsheryl** [comment]
After various attempts to get the setup working we tried doing a complete rebuild of the OS in one of the servers running the topology MySQL instances and discovery started working.

We tried to do a complete comparison of the working OS build with the one that did not work but we were not able to pin point what could be the problem - some OS setting(Does not appear to be any of the tcp settings as we suspected) or a library thing.

But we have completed a successful "proof of concept" run and with a ProxySQL front we have a functional application abstracted automatic failover solution with Orchestrator.

Thanks for your valuable time and extraordinary software!!
